tools/nvprefs/nvprefs.cpp

```diff
@@ -0,0 +1,227 @@
+
+      status = NvAPI_DRS_CreateApplication(session_handle, profile_handle, &application);
+      if (status != NVAPI_OK) return status;
+      session_modified = true;
+    }
+
+    NVDRS_SETTING setting = {};
+    setting.version = NVDRS_SETTING_VER1;
+    status = NvAPI_DRS_GetSetting(session_handle, profile_handle, PREFERRED_PSTATE_ID, &setting);
+
+    if (status != NVAPI_OK ||
+        setting.settingLocation != NVDRS_CURRENT_PROFILE_LOCATION ||
+        setting.u32CurrentValue != PREFERRED_PSTATE_PREFER_MAX) {  // Set power setting if needed
+      setting = {};
+      setting.version = NVDRS_SETTING_VER1;
+      setting.settingId = PREFERRED_PSTATE_ID;
+      setting.settingType = NVDRS_DWORD_TYPE;
+      setting.settingLocation = NVDRS_CURRENT_PROFILE_LOCATION;
+      setting.u32CurrentValue = PREFERRED_PSTATE_PREFER_MAX;
+
+      status = NvAPI_DRS_SetSetting(session_handle, profile_handle, &setting);
+      if (status != NVAPI_OK) return status;
+      session_modified = true;
+
+      std::cout << "Modified application power profile" << std::endl;
+    }
+    else {
+      std::cout << "Application power profile has already been modified" << std::endl;
+    }
+
+    return NVAPI_OK;
+  }
+
+  NvAPI_Status
+  check_and_modify_global_profile(NvDRSSessionHandle session_handle, bool &session_modified) {
+    NvAPI_Status status;
+
+    NvDRSProfileHandle profile_handle = 0;
+    status = NvAPI_DRS_GetBaseProfile(session_handle, &profile_handle);
+    if (status != NVAPI_OK) return status;
+
+    NVDRS_SETTING setting = {};
+    setting.version = NVDRS_SETTING_VER1;
+    status = NvAPI_DRS_GetSetting(session_handle, profile_handle, OGL_CPL_PREFER_DXPRESENT_ID, &setting);
+
+    if (status != NVAPI_OK ||
+        setting.u32CurrentValue != OGL_CPL_PREFER_DXPRESENT_PREFER_ENABLED) {  // Set OpenGL/Vulkan DXGI swapchain setting if needed
+      setting = {};
+      setting.version = NVDRS_SETTING_VER1;
+      setting.settingId = OGL_CPL_PREFER_DXPRESENT_ID;
+      setting.settingType = NVDRS_DWORD_TYPE;
+      setting.settingLocation = NVDRS_CURRENT_PROFILE_LOCATION;
+      setting.u32CurrentValue = OGL_CPL_PREFER_DXPRESENT_PREFER_ENABLED;
+
+      status = NvAPI_DRS_SetSetting(session_handle, profile_handle, &setting);
+      if (status != NVAPI_OK) return status;
+      session_modified = true;
+
+      std::cout << "Modified global OpenGL/Vulkan DXGI swapchain" << std::endl;
+    }
+    else {
+      std::cout << "No need to modify OpenGL/Vulkan DXGI swapchain" << std::endl;
+    }
+
+    return NVAPI_OK;
+  }
+
+  NvAPI_Status
+  restore_global_profile(NvDRSSessionHandle session_handle, bool &session_modified) {
+    NvAPI_Status status;
+
+    NvDRSProfileHandle profile_handle = 0;
+    status = NvAPI_DRS_GetBaseProfile(session_handle, &profile_handle);
+    if (status != NVAPI_OK) return status;
+
+    status = NvAPI_DRS_RestoreProfileDefaultSetting(session_handle, profile_handle, OGL_CPL_PREFER_DXPRESENT_ID);

```

Comment:
We probably want to remember that we changed this and only restore the previous setting if we changed it. Otherwise we might revert this setting to default after a user changed it manually.

Reply:
Will work until the first crash or otherwise improper system shutdown. Pretty much all systems will normalize to "enabled" sooner or later.

Reply:
I guess we can save it to file?

Reply:
Yeah, a file or a registry key seems reasonable.

Reply:
Still, one little mistake and we pollute the settings with false "enabled", many ways it can go wrong. Resetting to default is simple, and should be reasonably safe to merge this close to release. I believe anything more complex needs more time in nightly.

We can reset to default this release, and leave the complex logic for the next. Or we can simply move the whole thing to the next. Thoughts?

Reply:
*Still*, we can similarly pollute the system during uninstallation, if it doesn't terminate sunshine.exe gracefully. Let's move it to the next release after all, too many potential problems.

Reply:
The things to keep track of (so far)
 - use registry, to persist across reinstallations, especially portable
 - call restore from the uninstaller, in case sunshine didn't exit gracefully
 - save "using default" as separate value, in case nvidia decides to change it

Reply:
Some potential useful things to keep in mind:

* nvapi drs api offers [`NvAPI_DRS_SaveSettingsToFile()`](https://docs.nvidia.com/gameworks/content/gameworkslibrary/coresdk/nvapi/group__drsapi.html#gaf2f69fee6df6510c8cb5a5468b0e0b05), this call makes a full copy of the settings db to a file. The backup file can be loaded and navigated the same way as the actual settings.
* I agree that restoring previous settings is important. Nvidia driver settings use inheritance so this is not so trivial. 
  * In our case, we can probably assume that sunshine will have no "bundled" app profile (an app profile bundled with the driver). This allows us to only use `.settingLocation != NVDRS_CURRENT_PROFILE_LOCATION` to determine if a setting has to be set to a previously user value or reset. (In a general case, we would also need to check `.isCurrentPredefined`).
  * _Getting_ the current setting value can fail if the setting is inherited. Just a reminder.
*  That said, restoring a custom setting for the sunshine profile is probably unnecessary. The global setting must be restored. With the assumption above, the restore process uses the same set-or-reset logic.
* As you wrote above, guaranteeing restoration is complicated. Avoiding global profile tampering altogether might simply be better? The user can use the tool as a do-undo command.